gh-79125: Update _dirnameW to accept long path names

jopamer opened the pull request:
The fix for issue 32557 updated `os__getdiskusage_impl` to use `_dirnameW` for obtaining the parent directory of a file. This would cause a regression if the path exceeded 260 characters, since `_dirnameW` currently returns -1 if given a path >= `MAX_PATH` in length.

As suggested in the issue's comments, `_dirnameW` should be updated to use `PathCchRemoveFileSpec` when available (on Win8.1 or greater) to avoid throwing an unnecessary error on a long path.

Note:
If `PathCchRemoveFileSpec` isn't available, we can call through `PathRemoveFileSpecW`, which is otherwise deprecated. What's interesting there is that while the docs say it expects a buffer of size `MAX_PATH`, analysis of the function shows that it doesn't make assumptions about the size of the path other than it's less than 32k characters in length. It calls through `PathCchRemoveFileSpec` under the hood on Win8 and greater, passing in 0x8000h as the `cch` argument. `PathCchRemoveFileSpec` then scans through the path for '\' via `wcschr`, stops when it hits the last one and inserts a NULL. (Analysis of `PathRemoveFileSpecW` on a Win7 VM shows that it does basically the same thing, and is also resilient to paths greater than `MAX_PATH` in length.)

<!-- issue-number: [bpo-34944](https://www.bugs.python.org/issue34944) -->
https://bugs.python.org/issue34944
<!-- /issue-number -->


<!-- gh-issue-number: gh-79125 -->
* Issue: gh-79125
<!-- /gh-issue-number -->

jopamer committed:
Update _dirnameW to use PathCchRemoveFileSpec when available

Calling through PathCchRemoveFileSpec will allow _dirnameW to trim paths longer than MAX_PATH in length.

jopamer committed:
Remove dependency on MAX_PATH in _dirnameW and its users

zooba commented on Modules/posixmodule.c:
```diff
@@ -7657,24 +7658,36 @@ check_CreateSymbolicLink(void)
-            break;
+    if (_PPathCchRemoveFileSpec) {
+        if (FAILED(_PPathCchRemoveFileSpec(path, length))) {
+            Py_FatalError("Error removing path element in _dirnameW");
```
I'm not excited about adding a fatal error in the midst of a Python call, especially since we have an error code here that could be returned instead (inside `getpathp.c` is different, because we aren't initialized enough to raise an exception at that stage).

Setting an error here and returning non-zero should work fine. I _believe_ all calls to this are already holding the GIL, but that would be the only issue.

jopamer replied:
Thanks - I wasn't aware of the difference there. I'll update the diff.

zooba commented on Modules/posixmodule.c:
```diff
@@ -7657,24 +7658,36 @@ check_CreateSymbolicLink(void)
+    size_t length = wcslen(path);
+
+    if (!_PPathCchRemoveFileSpec_Initialized) {
+        HMODULE pathapi = LoadLibraryW(L"api-ms-win-core-path-l1-1-0.dll");
```
`LoadLibrary` calls should really be outside the GIL, as they can trigger a lot of work.

If we require that the GIL is held on entry to this function, then wrapping the entire contents of this initialization block in `Py_BEGIN_ALLOW_THREADS` should be sufficient. There's likely more lost that gained if we go too fine-grained with it.

jopamer replied:
Ok, cool - I'll make that change and see what happens.

jopamer committed:
Fold in code review feedback

* Revert _dirnameW back to returning 0 on success, -1 on failure.
* Release the GIL before calling out to LoadLibraryW.

jopamer committed:
Merge branch 'master' into bpo-34944

jopamer committed:
Fix two bits of diff mistakes

jopamer committed:
Roll back changes to _check_dirW

_joinW presumes a buffer length of MAX_PATH, so given that I've changed the return type of _dirnameW back to int, I'll revert my changes for now.

jopamer committed:
Make sure os_symlink_impl has the GIL when it invokes _check_dirW

Also, update _joinW to not fix buffer sizes to MAX_PATH.

Commit:
Add NEWS.d entry
